### PR TITLE
Armstrong

### PR DESCRIPTION
**Pull request by sedgewickmm18**
Updates to the anomaly functions

### CODE REVIEW

File: iotfunctions/anomaly.py [hunk 1]
```diff
@@ -204,19 +152,53 @@ def execute(self, df):
...
+                    # compute the elliptic envelope to exploit Minimum Covariance Determinant estimates
+                    #    standardizing
+                    lowETS = (lowETS - lowETS.mean())/lowETS.std(ddof=0)
+                    highETS = (highETS - highETS.mean())/highETS.std(ddof=0)
+
+                    twoDimETS = np.vstack((lowETS, highETS)).T
+                    logger.info('lowETS: ' + str(lowETS) + ', highETS:' + str(highETS) + 'input' + str(twoDimETS))
+
+                    # inliers have a score of 1, outliers -1, and 0 indicates an issue with the data
+                    dfe[self.output_item] = 0.0002
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: What are these hardcoded values?

File: iotfunctions/anomaly.py [hunk 2]
```diff
@@ -204,19 +152,53 @@ def execute(self, df):
...
+                    ets_zscore = ellEnv.decision_function(twoDimETS, raw_values=True).copy()
+
+                    # compute zscore over the energy
+                    #ets_zscore = np.abs((ETS - ETS.mean())/ETS.std(ddof=0))
+                    logger.info('Spectral z-score max: ' + str(ets_zscore.max()))
+
+                    # length of timesTS, ETS and ets_zscore is smaller than half the original
+                    #   extend it to cover the full original length 
+                    dfe[self.output_item] = 0.0005
+                    Linear = sp.interpolate.interp1d(timesTS, ets_zscore, kind='linear', fill_value='extrapolate')
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: Good to follow variable naming convention in python. example: variable_test_value, varable_example. Only constant will be capital letters.

File: iotfunctions/anomaly.py [hunk 3]
```diff
@@ -229,7 +211,7 @@ def execute(self, df):
                 elif self.output_item in dfe_orig:
                     zScoreII = dfe_orig[self.output_item].to_numpy()
                 else:
-                    #print (dfe_orig.head(2))
+                    print (dfe_orig.head(2))
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: Please use a logger instead of a print.

File: iotfunctions/anomaly.py [hunk 4]
```diff
@@ -272,7 +257,7 @@ class SpectralAnomalyScore(BaseTransformer):
     '''
     def __init__(self, input_item, windowsize, output_item):
         super().__init__()
-        logger.debug(input_item)
+        logger.info(input_item)
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: It is better to place debug level log for finding out input.

File: iotfunctions/anomaly.py [hunk 5]
```diff
@@ -293,7 +278,7 @@ def execute(self, df):
 
         df_copy = df.copy()
         entities = np.unique(df.index.levels[0])
-        logger.debug(str(entities))
+        logger.info(str(entities))
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: same as above

File: iotfunctions/anomaly.py [hunk 6]
```diff
@@ -309,9 +294,7 @@ def execute(self, df):
             dfe_orig = dfe_orig.reset_index(level=[0]).sort_index()
 
             # minimal time delta for merging
-            mindelta = dfe_orig.index.to_series().diff().min()
-            if mindelta == dt.timedelta(seconds = 0) or pd.isnull(mindelta):
-                mindelta = pd.Timedelta('5 seconds')
+            mindelta = minDelta(dfe_orig)
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: naming convensions. even for methods. example: metod_name()

File: iotfunctions/anomaly.py [hunk 7]
```diff
@@ -322,11 +305,14 @@ def execute(self, df):
             # one dimensional time series - named temperature for catchyness
             temperature = dfe[[self.input_item]].fillna(0).to_numpy().reshape(-1,)
 
-            logger.debug('Spectral: ' + str(entity) + ', ' + str(self.input_item) + ', ' + str(self.windowsize) + ', ' +
+            logger.info('Spectral: ' + str(entity) + ', ' + str(self.input_item) + ', ' + str(self.windowsize) + ', ' +
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: debug over info. Add info only if  you think the user may make some mistake in this stage.

File: iotfunctions/anomaly.py [hunk 8]
```diff
@@ -338,19 +324,54 @@ def execute(self, df):
                 freqsTS = freqsTS * freqsTSb
                 freqsTS[freqsTS == 0] = 1 / self.windowsize
 
+                highfreqsTS = freqsTS.copy()
+                lowfreqsTS = freqsTS.copy()
+                highfreqsTS[highfreqsTS <= 0.3] = 0
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: Save it in a constant ( 0.3 )

File: iotfunctions/anomaly.py [hunk 9]
```diff
@@ -363,7 +384,7 @@ def execute(self, df):
                 elif self.output_item in dfe_orig:
                     zScoreII = dfe_orig[self.output_item].to_numpy()
                 else:
-                    #print (dfe_orig.head(2))
+                    print (dfe_orig.head(2))
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: logger over print

File: iotfunctions/anomaly.py (hunk 1 above)
Comment (sedgewickmm18) [reply iotfunctions/anomaly.py]: minimal values close to zero to act as error return codes

File: iotfunctions/anomaly.py [hunk 10]
```diff
@@ -150,7 +100,7 @@ def execute(self, df):
 
         df_copy = df.copy()
         entities = np.unique(df.index.levels[0])
-        logger.debug(str(entities))
+        logger.info(str(entities))
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: debug over info.

File: iotfunctions/anomaly.py [hunk 11]
```diff
@@ -167,32 +117,30 @@ def execute(self, df):
...
-            if temperature.size > self.windowsize:
-                logger.debug(str(temperature.size) + str(self.windowsize))
+            if temperature.size <= self.windowsize:
+                logger.info(str(temperature.size) + ' <= ' + str(self.windowsize))
+                df_copy.loc[[entity]] = 0.0001
+            else:
+                logger.info(str(temperature.size) + str(self.windowsize))
                 # Fourier transform:
                 #   frequency, time, spectral density
                 freqsTS, timesTS, SxTS = signal.spectrogram(temperature, fs = self.frame_rate, window = 'hanning',
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: name variable accordingly(  frequency, time, spectral density)

File: iotfunctions/anomaly.py (hunk 2 above)
Comment (sedgewickmm18) [reply iotfunctions/anomaly.py]: changed the name to linear_interpolate

File: iotfunctions/anomaly.py [hunk 12]
```diff
@@ -204,19 +152,53 @@ def execute(self, df):
...
+                lowfreqsTS[lowfreqsTS > 0.3] = 0
 
-                # compute zscore over the energy
-                ets_zscore = (ETS - ETS.mean())/ETS.std(ddof=0)
-                logger.debug('NoData z-score max: ' + str(ets_zscore.max()))
+                # Compute energy = frequency * spectral density over time in decibel
+                try:
+                    lowETS = np.log10(np.dot(SxTS.T, lowfreqsTS))
+                    highETS = np.log10(np.dot(SxTS.T, highfreqsTS))
+                    ETS = np.log10(np.dot(SxTS.T, freqsTS))
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: what is ETS ? is it constant?

File: iotfunctions/anomaly.py (hunk 3 above)
Comment (sedgewickmm18) [reply iotfunctions/anomaly.py]: removed all occurrences of 'print'

File: iotfunctions/anomaly.py [hunk 13]
```diff
@@ -204,19 +152,53 @@ def execute(self, df):
...
+                    logger.info('lowETS: ' + str(lowETS) + ', highETS:' + str(highETS) + 'input' + str(twoDimETS))
+
+                    # inliers have a score of 1, outliers -1, and 0 indicates an issue with the data
+                    dfe[self.output_item] = 0.0002
+                    ellEnv = EllipticEnvelope(random_state=0)
+
+                    dfe[self.output_item] = 0.0003
+                    ellEnv.fit(twoDimETS)
+                    #ets_zscore = ellEnv.predict(twoDimETS)
+                    dfe[self.output_item] = 0.0004
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: May I know why this hard coding?  Comment may help to understand the situation

File: iotfunctions/anomaly.py (hunk 4 above)
Comment (sedgewickmm18) [reply iotfunctions/anomaly.py]: done

File: iotfunctions/anomaly.py (hunk 6 above)
Comment (sedgewickmm18) [reply iotfunctions/anomaly.py]: so no camelCase ?

Comment (sedgewickmm18) [reply iotfunctions/anomaly.py]: renamed

File: iotfunctions/anomaly.py (hunk 8 above)
Comment (sedgewickmm18) [reply iotfunctions/anomaly.py]: Using FrequencySplit constant now

File: iotfunctions/anomaly.py (hunk 11 above)
Comment (sedgewickmm18) [reply iotfunctions/anomaly.py]: it's now 
frequency_temperature, time_series_temperature, spectral_density_temperature

File: iotfunctions/anomaly.py (hunk 12 above)
Comment (sedgewickmm18) [reply iotfunctions/anomaly.py]: renamed to signal_energy  -  from E(nergy)T(emperature)S(ignal)

File: iotfunctions/anomaly.py (hunk 13 above)
Comment (sedgewickmm18) [reply iotfunctions/anomaly.py]: same hardcoding for error numbers close to zero

File: iotfunctions/anomaly.py [hunk 14]
```diff
@@ -325,32 +310,70 @@ def execute(self, df):
...
+
+                highfrequency_temperature = frequency_temperature.copy()
+                lowfrequency_temperature = frequency_temperature.copy()
+                highfrequency_temperature[highfrequency_temperature <= FrequencySplit] = 0
+                lowfrequency_temperature[lowfrequency_temperature > FrequencySplit] = 0
 
                 # Compute energy = frequency * spectral density over time in decibel
-                ETS = np.log10(np.dot(SxTS.T, freqsTS))
+                try:
+                    lowsignal_energy = np.log10(np.dot(spectral_density_temperature.T, lowfrequency_temperature))
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: low_signal_energy

File: iotfunctions/anomaly.py [hunk 15]
```diff
@@ -325,32 +310,70 @@ def execute(self, df):
...
+                    highsignal_energy = np.log10(np.dot(spectral_density_temperature.T, highfrequency_temperature))
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: high_frequency_temperature

File: iotfunctions/anomaly.py [hunk 16]
```diff
@@ -325,32 +310,70 @@ def execute(self, df):
...
+                    # compute the elliptic envelope to exploit Minimum Covariance Determinant estimates
+                    #    standardizing
+                    lowsignal_energy = (lowsignal_energy - lowsignal_energy.mean())/lowsignal_energy.std(ddof=0)
+                    highsignal_energy = (highsignal_energy - highsignal_energy.mean())/highsignal_energy.std(ddof=0)
 
-                # length of timesTS, ETS and ets_zscore is smaller than half the original
-                #   extend it to cover the full original length 
-                Linear = sp.interpolate.interp1d(timesTS, ets_zscore, kind='linear', fill_value='extrapolate')
-                zscoreI = Linear(np.arange(0, temperature.size, 1))
+                    twoDimsignal_energy = np.vstack((lowsignal_energy, highsignal_energy)).T
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: two_dimsignal_energy

File: iotfunctions/anomaly.py [hunk 17]
```diff
@@ -325,32 +310,70 @@ def execute(self, df):
...
+                    logger.debug('lowsignal_energy: ' + str(lowsignal_energy) + ', highsignal_energy:' + str(highsignal_energy) + 'input' + str(twoDimsignal_energy))
+
+                    # inliers have a score of 1, outliers -1, and 0 indicates an issue with the data
+                    dfe[self.output_item] = 0.0002
+                    ellEnv = EllipticEnvelope(random_state=0)
```
Comment (herleraja) [inline iotfunctions/anomaly.py]: elliptic_envelope ?

Comment (sedgewickmm18) [reply iotfunctions/anomaly.py]: the name of the method, the assumption is that data is normally distributed. The algo finds a suitable coordinate system for the majority of the data points and looks for the elliptic boundary (with respect to this new coordinate system). Data points outside of the ellipse are outliers

**Pull request merged**